### ISSUE TITLE
Misindented line before comment

### ISSUE THREAD

**Issue by hugovk** [asottile/pyupgrade#358]
pyupgrade==2.7.3

Running `pyupgrade 1.py --py3-plus`

# With comment

## Before

```python
import six


class HTTPConnection(_HTTPConnection, object):

    def __init__(self, *args, **kw):
        if not six.PY2:
            kw.pop("strict", None)

        # Comment.
        pass
```

## After

Misindented `kw.pop`:

```python
import six


class HTTPConnection(_HTTPConnection):

    def __init__(self, *args, **kw):
            kw.pop("strict", None)

        # Comment.
        pass
```

# Without comment

## Before

```python
import six


class HTTPConnection(_HTTPConnection, object):

    def __init__(self, *args, **kw):
        if not six.PY2:
            kw.pop("strict", None)

        pass
```

## After

Indent is good.

```python
import six


class HTTPConnection(_HTTPConnection):

    def __init__(self, *args, **kw):
        kw.pop("strict", None)

        pass
```

**Comment by asottile** [comment]
ah this is the deden hoisting problem again:

```
(Pdb) self
Block(start=32, colon=38, block=40, end=56, line=False)
(Pdb) pp tuple(enumerate(tokens))
((0, Token(name='NAME', src='import', line=1, utf8_byte_offset=0)),
 (1, Token(name='UNIMPORTANT_WS', src=' ', line=None, utf8_byte_offset=None)),
 (2, Token(name='NAME', src='six', line=1, utf8_byte_offset=7)),
 (3, Token(name='NEWLINE', src='\n', line=1, utf8_byte_offset=10)),
 (4, Token(name='NL', src='\n', line=2, utf8_byte_offset=0)),
 (5, Token(name='NL', src='\n', line=3, utf8_byte_offset=0)),
 (6, Token(name='NAME', src='class', line=4, utf8_byte_offset=0)),
 (7, Token(name='UNIMPORTANT_WS', src=' ', line=None, utf8_byte_offset=None)),
 (8, Token(name='NAME', src='HTTPConnection', line=4, utf8_byte_offset=6)),
 (9, Token(name='OP', src='(', line=4, utf8_byte_offset=20)),
 (10, Token(name='NAME', src='_HTTPConnection', line=4, utf8_byte_offset=21)),
 (11, Token(name='OP', src=')', line=4, utf8_byte_offset=36)),
 (12, Token(name='OP', src=':', line=4, utf8_byte_offset=37)),
 (13, Token(name='NEWLINE', src='\n', line=4, utf8_byte_offset=38)),
 (14, Token(name='NL', src='\n', line=5, utf8_byte_offset=0)),
 (15, Token(name='INDENT', src='    ', line=6, utf8_byte_offset=0)),
 (16, Token(name='NAME', src='def', line=6, utf8_byte_offset=4)),
 (17, Token(name='UNIMPORTANT_WS', src=' ', line=None, utf8_byte_offset=None)),
 (18, Token(name='NAME', src='__init__', line=6, utf8_byte_offset=8)),
 (19, Token(name='OP', src='(', line=6, utf8_byte_offset=16)),
 (20, Token(name='NAME', src='self', line=6, utf8_byte_offset=17)),
 (21, Token(name='OP', src=',', line=6, utf8_byte_offset=21)),
 (22, Token(name='UNIMPORTANT_WS', src=' ', line=None, utf8_byte_offset=None)),
 (23, Token(name='OP', src='*', line=6, utf8_byte_offset=23)),
 (24, Token(name='NAME', src='args', line=6, utf8_byte_offset=24)),
 (25, Token(name='OP', src=',', line=6, utf8_byte_offset=28)),
 (26, Token(name='UNIMPORTANT_WS', src=' ', line=None, utf8_byte_offset=None)),
 (27, Token(name='OP', src='**', line=6, utf8_byte_offset=30)),
 (28, Token(name='NAME', src='kw', line=6, utf8_byte_offset=32)),
 (29, Token(name='OP', src=')', line=6, utf8_byte_offset=34)),
 (30, Token(name='OP', src=':', line=6, utf8_byte_offset=35)),
 (31, Token(name='NEWLINE', src='\n', line=6, utf8_byte_offset=36)),
 (32, Token(name='INDENT', src='        ', line=7, utf8_byte_offset=0)),
 (33, Token(name='NAME', src='if', line=7, utf8_byte_offset=8)),
 (34, Token(name='UNIMPORTANT_WS', src=' ', line=None, utf8_byte_offset=None)),
 (35, Token(name='NAME', src='six', line=7, utf8_byte_offset=11)),
 (36, Token(name='OP', src='.', line=7, utf8_byte_offset=14)),
 (37, Token(name='NAME', src='PY3', line=7, utf8_byte_offset=15)),
 (38, Token(name='OP', src=':', line=7, utf8_byte_offset=18)),
 (39, Token(name='NEWLINE', src='\n', line=7, utf8_byte_offset=19)),
 (40, Token(name='INDENT', src='            ', line=8, utf8_byte_offset=0)),
 (41, Token(name='NAME', src='kw', line=8, utf8_byte_offset=12)),
 (42, Token(name='OP', src='.', line=8, utf8_byte_offset=14)),
 (43, Token(name='NAME', src='pop', line=8, utf8_byte_offset=15)),
 (44, Token(name='OP', src='(', line=8, utf8_byte_offset=18)),
 (45, Token(name='STRING', src='"strict"', line=8, utf8_byte_offset=19)),
 (46, Token(name='OP', src=',', line=8, utf8_byte_offset=27)),
 (47, Token(name='UNIMPORTANT_WS', src=' ', line=None, utf8_byte_offset=None)),
 (48, Token(name='NAME', src='None', line=8, utf8_byte_offset=29)),
 (49, Token(name='OP', src=')', line=8, utf8_byte_offset=33)),
 (50, Token(name='NEWLINE', src='\n', line=8, utf8_byte_offset=34)),
 (51, Token(name='NL', src='\n', line=9, utf8_byte_offset=0)),
 (52,
  Token(name='UNIMPORTANT_WS', src='        ', line=None, utf8_byte_offset=None)),
 (53, Token(name='COMMENT', src='# Comment.', line=10, utf8_byte_offset=8)),
 (54, Token(name='NL', src='\n', line=10, utf8_byte_offset=18)),
 (55, Token(name='DEDENT', src='', line=11, utf8_byte_offset=8)),
 (56,
  Token(name='UNIMPORTANT_WS', src='        ', line=None, utf8_byte_offset=None)),
 (57, Token(name='NAME', src='pass', line=11, utf8_byte_offset=8)),
 (58, Token(name='NEWLINE', src='\n', line=11, utf8_byte_offset=12)),
 (59, Token(name='DEDENT', src='', line=12, utf8_byte_offset=0)),
 (60, Token(name='DEDENT', src='', line=12, utf8_byte_offset=0)),
 (61, Token(name='ENDMARKER', src='', line=12, utf8_byte_offset=0)))


basically that `DEDENT` token@55 is in the wrong spot, it should be between 51 and 52 probably

there's code [here](https://github.com/asottile/pyupgrade/blob/20d5c9ef3ca920c6332f2e0605fd0013225640f4/pyupgrade.py#L1745-L1757) which tries to fix it to be there, but misses this case

**Comment by int3l** [comment]
Interestingly enough, this examples raises `IndentationError`, when I try to test it.

**Comment by asottile** [comment]
this makes sense due to the plugin reshfufling

**Comment by int3l** [comment]
```python
def _fixup_dedent_tokens(tokens: List[Token]) -> None:
    """For whatever reason the DEDENT / UNIMPORTANT_WS tokens are misordered

    | if True:
    |     if True:
    |         pass
    |     else:
    |^    ^- DEDENT
    |+----UNIMPORTANT_WS
    """
    comment_idx = 0
    for i, token in enumerate(tokens):
        if token.name == UNIMPORTANT_WS:
            if tokens[i + 1].name == 'COMMENT':
                comment_idx = i
            elif tokens[i + 1].name == 'DEDENT':
                if comment_idx:
                    tokens.insert(comment_idx, tokens.pop(i + 1))
                    comment_idx = 0
                else:
                    tokens[i], tokens[i + 1] = tokens[i + 1], tokens[i]
```

Is this even acceptable? I don't like the pop/insert solution.

**Comment by asottile** [comment]
there's already the beginnings of a patch in #362 -- I think the correct fix is to abandon this fixup method and instead adjust how the `Block` parsing works (treating comment-only / whitespace-only lines as not a minimum indent and not reducing their indent level below the minimum of the block) -- because this sort of thing is valid:

```python
if True:
#comment
    code
```

**Comment by int3l** [comment]
Oh, nice.
Another test case that fails is `white space after comment`:
```python
if not six.PY2:
    ...
# comment
    
```

**Comment by asottile** [comment]
via #362
